src/script/src/python/coprocessor.rs

```diff
@@ -318,10 +346,83 @@ pub fn exec_coprocessor(script: &str, rb: &RecordBatch) -> Result<RecordBatch> {
     // 1. parse the script and check if it's only a function with `@coprocessor` decorator, and get `args` and `returns`,
     // 2. also check for exist of `args` in `rb`, if not found, return error
     // TODO(discord9): cache the result of parse_copr
-    let copr = parse::parse_and_compile_copr(script)?;
+    let copr = parse::parse_and_compile_copr(script, None)?;
     exec_parsed(&copr, rb)
 }
 
+#[pyclass(module = false, name = "query_engine")]
+#[derive(Debug, PyPayload)]
+pub struct PyQueryEngine {
+    inner: QueryEngineWeakRef,
+}
+
+#[pyclass]
+impl PyQueryEngine {
+    // TODO(discord9): find a better way to call sql query api, now we don't if we are in async contex or not
+    #[pymethod]
+    fn sql(&self, s: String, vm: &VirtualMachine) -> PyResult<PyListRef> {
+        let query = self.inner.0.upgrade();
+        let handle = std::thread::spawn(move || -> std::result::Result<RecordBatches, String> {
+            if let Some(engine) = query {
+                let stmt = QueryLanguageParser::parse_sql(s.as_str()).map_err(|e| e.to_string())?;
+                let plan = engine
+                    .statement_to_plan(stmt, Default::default())
+                    .map_err(|e| e.to_string())?;
+                // To prevent the error of nested creating Runtime, because we don't know if we are being called from a async context
+                let rt = tokio::runtime::Runtime::new().map_err(|e| e.to_string())?;
```

Comment:
You can `unwrap_or_else()` here, using [`Handle::current`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.current) to retrieve the runtime handle if there is one.